### PR TITLE
relax provider versions for terraform 0.12 readiness

### PR DESCRIPTION
**Pull request by isen-ng**
If provider versions are fixed inside modules, parent terraform scripts cannot specify a different version as it will cause version conflicts.

Instead of removing provider versions all together, it might be beneficial to relax the versioning from a strict version to a greater than version. This hopefully sends the message that this modules will work with a provider version of the at least specified version.

Context:
I'm trying to prepare my terraform to be 0.12 ready and they recommend that I upgrade some providers. However I cannot upgrade them since their versions are fixed in this child module.

### CODE REVIEW

**Comment by sshvetsov** [comment]
@isen-ng what would be your expectations from the maintainers when upstream provider introduces breaking changes that impact this module? If we set only minimum provider version as a requirement a new major version can potentially break existing functionality. I think a more safe approach would be to set the tested version of the provider as a minimum and use "up to" constraint for the major version, i.e. `version = "~>1.0"` for "null" provider and `version = "~>1.2"` for "random" provider.

**Comment by sshvetsov** [comment]
I'm also wondering if there's still a need to directly use "random" provider at all in this module. Perhaps we can replace random_id resource with call out to [aws-resource-naming](https://github.com/traveloka/terraform-aws-resource-naming) module instead?

**Comment by isen-ng** [comment]
> I think a more safe approach would be to set the tested version of the provider as a minimum and use "up to" constraint for the major version, i.e. version = "~>1.0" for "null" provider and version = "~>1.2" for "random" provider.

I think you make a very good point here. 

I'm my use case (not urgent use case, but a nice to have use case), I'm trying to get my modules to be 0.12 ready. Not upgrading but getting providers and making non-breaking changes to ease the transition to 0.12 when the time comes. One of them is to upgrade providers to versions that support 0.12.

Currently when I install terraform 0.11.14 and run `terraform 0.12checklist`, it outputs that some providers needs to be upgraded. For example, the random provider should be "~> 2.1.2".

But, like you said, I think upgrading providers blindly might be dangerous for downstream projects

**Comment by isen-ng** [comment]
> I'm also wondering if there's still a need to directly use "random" provider at all in this module. Perhaps we can replace random_id resource with call out to aws-resource-naming module instead?

this is a good idea. The less repetitions, the less to maintain.

**Comment by isen-ng** [comment]
I want to be able to be 0.12 ready but how should I go about requesting this change considering the impact it might bring?

**Comment by sshvetsov** [comment]
If your goal is move to terraform v0.12, then most likely all modules will require a major version bump as none were written for or tested with v0.12. That's a pretty major undertaking and we will need some help testing the modules with new version.

**Comment by isen-ng** [comment]
The end goal is the move to 0.12 but it takes a lot to move (and probably some time for them to fix new bugs).
I'm now currently thinking about breaking things up and only be 0.12-ready, still continuing using 0.11.14.

I am open to helping to write code and test things out.

**Comment by sshvetsov** [comment]
I think we should split this into two parts:
#1 relaxing the provider support beyond minimum version, but only up to next major version of the provider (this should be useful for people still using terraform v0.11)
#2 creating new major versions with support for terraform v0.12 (this might require a separate branch or even repo and a lot more discussion)

**Comment by isen-ng** [comment]
> #1 relaxing the provider support beyond minimum version, but only up to next major version of the provider (this should be useful for people still using terraform v0.11)

I can send PRs for the modules that I am looking at now, namely
aws-resource-naming,
aws-iam-role, and 
aws-vpc

File: modules/external/main.tf [hunk 1]
```diff
@@ -1,5 +1,5 @@
 provider "null" {
-  version = "1.0.0"
+  version = ">= 1.0.0"
```
Comment (sshvetsov) [inline modules/external/main.tf]: Let's change this to `version = "~> 1.0.0"` instead. This will allow us to bump only the patch version, once testing and changelog diving confirms there are no breaking changes between in "null" provider between 1.0 and 2.0

File: modules/instance/main.tf [hunk 2]
```diff
@@ -1,5 +1,5 @@
 provider "random" {
-  version = "1.2.0"
+  version = ">= 1.2.0"
```
Comment (sshvetsov) [inline modules/instance/main.tf]: Let's change this to `version = "~> 1.2.0"` instead. This will allow us to bump only the patch version, once testing and changelog diving confirms there are no breaking changes between in "random" provider between 1.2 and 2.0

File: modules/service/main.tf [hunk 3]
```diff
@@ -1,5 +1,5 @@
 provider "random" {
-  version = "1.2.0"
+  version = ">= 1.2.0"
```
Comment (sshvetsov) [inline modules/service/main.tf]: Let's change this to `version = "~> 1.2.0"` instead. This will allow us to bump only the patch version, once testing and changelog diving confirms there are no breaking changes between in "random" provider between 1.2 and 2.0

**Comment by isen-ng** [comment]
> creating new major versions with support for terraform v0.12 (this might require a separate branch or even repo and a lot more discussion)

If my help is needed for this, I'll be glad to help.

**Comment by isen-ng** [comment]
Closing as this PR is no longer valid

File: modules/external/main.tf [hunk 4]
```diff
@@ -1,5 +1,5 @@
 provider "null" {
-  version = "1.0.0"
+  version = "~> 1.0.0"
```
Comment (rafikurnia) [inline modules/external/main.tf]: `~> 1.0.0` means `>= 1.0.0, < 1.1.0`
Null provider supports Terraform 0.12 from version [2.0.0](https://github.com/terraform-providers/terraform-provider-null/blob/master/CHANGELOG.md#200-january-18-2019)

I suggest to change it to `>= 1.0.0, < 3.0.0`

File: modules/instance/main.tf [hunk 5]
```diff
@@ -1,5 +1,5 @@
 provider "random" {
-  version = "1.2.0"
+  version = "~> 1.2.0"
```
Comment (rafikurnia) [inline modules/instance/main.tf]: `~> 1.2.0` means `>= 1.2.0, < 1.3.0`
Random provider supports Terraform 0.12 from version [2.1.0](https://github.com/terraform-providers/terraform-provider-random/blob/master/CHANGELOG.md#210-march-20-2019)

I suggest to change it to `>= 1.2.0, < 3.0.0``~> 1.2.0` means `>= 1.2.0, < 1.3.0`
Random provider supports Terraform 0.12 from version [2.1.0](https://github.com/terraform-providers/terraform-provider-random/blob/master/CHANGELOG.md#210-march-20-2019)

I suggest to change it to `>= 1.2.0, < 3.0.0`

File: modules/service/main.tf [hunk 6]
```diff
@@ -1,5 +1,5 @@
 provider "random" {
-  version = "1.2.0"
+  version = "~> 1.2.0"
```
Comment (rafikurnia) [inline modules/service/main.tf]: `~> 1.2.0` means `>= 1.2.0, < 1.3.0`
Random provider supports Terraform 0.12 from version [2.1.0](https://github.com/terraform-providers/terraform-provider-random/blob/master/CHANGELOG.md#210-march-20-2019)

I suggest to change it to `>= 1.2.0, < 3.0.0`
